File: pipeline/highlighter_pipeline/storage.py
import mimetypes
import os
from pathlib import Path

from .defaults import DEFAULT_AWS_REGION, DEFAULT_S3_BUCKET

_CONTENT_TYPES = {
    ".ts": "video/mp2t",
    ".wav": "audio/wav",
    ".mp4": "video/mp4",
}


class S3Storage:
    def __init__(self, bucket: str, region: str) -> None:
        import boto3  # deferred so S3-disabled runs don't need it installed

        self.bucket = bucket
        self.region = region
        self._client = boto3.client("s3", region_name=region)

    def upload_file(self, path: Path, key: str) -> str:
        content_type = (
            _CONTENT_TYPES.get(path.suffix.lower())
            or mimetypes.guess_type(path.name)[0]
            or "application/octet-stream"
        )
        self._client.upload_file(
            str(path), self.bucket, key, ExtraArgs={"ContentType": content_type}
        )
        return self.url_for(key)

    def url_for(self, key: str) -> str:
        return f"https://{self.bucket}.s3.{self.region}.amazonaws.com/{key}"

    def download_file(self, key: str, path: Path) -> Path:
        path.parent.mkdir(parents=True, exist_ok=True)
        self._client.download_file(self.bucket, key, str(path))
        return path

    def delete_object(self, key: str) -> None:
        self._client.delete_object(Bucket=self.bucket, Key=key)

    def delete_prefix(self, prefix: str) -> int:
        """Delete every object under a prefix. Returns the number deleted."""
        deleted = 0
        paginator = self._client.get_paginator("list_objects_v2")
        for page in paginator.paginate(Bucket=self.bucket, Prefix=prefix.rstrip("/") + "/"):
            keys = [{"Key": item["Key"]} for item in page.get("Contents", [])]
            if keys:
                self._client.delete_objects(Bucket=self.bucket, Delete={"Objects": keys})
                deleted += len(keys)
        return deleted


def storage_from_env() -> S3Storage | None:
    """S3 storage from S3_BUCKET/AWS_REGION. Empty S3_BUCKET disables uploads."""
    bucket = os.environ.get("S3_BUCKET", DEFAULT_S3_BUCKET)
    if not bucket:
        return None
    return S3Storage(bucket, os.environ.get("AWS_REGION", DEFAULT_AWS_REGION))

File: pipeline-dotnet/src/Highlighter.Pipeline/Storage.cs
using Amazon;
using Amazon.S3;
using Amazon.S3.Model;

namespace Highlighter.Pipeline;

/// <summary>Port of highlighter_pipeline/storage.py (boto3 → AWSSDK.S3; the
/// default credential chain behaves the same way).</summary>
public class S3Storage
{
    private static readonly Dictionary<string, string> ContentTypes = new(StringComparer.OrdinalIgnoreCase)
    {
        [".ts"] = "video/mp2t",
        [".wav"] = "audio/wav",
        [".mp4"] = "video/mp4",
    };

    public string Bucket { get; }
    public string Region { get; }

    private readonly AmazonS3Client _client;

    public S3Storage(string bucket, string region)
    {
        Bucket = bucket;
        Region = region;
        _client = new AmazonS3Client(RegionEndpoint.GetBySystemName(region));
    }

    public string UploadFile(string path, string key)
    {
        var contentType = ContentTypes.TryGetValue(Path.GetExtension(path), out var mapped)
            ? mapped
            : MimeTypes.Guess(path) ?? "application/octet-stream";
        var request = new PutObjectRequest
        {
            BucketName = Bucket,
            Key = key,
            FilePath = path,
            ContentType = contentType,
        };
        _client.PutObjectAsync(request).GetAwaiter().GetResult();
        return UrlFor(key);
    }

    public string UrlFor(string key) => $"https://{Bucket}.s3.{Region}.amazonaws.com/{key}";

    public string DownloadFile(string key, string path)
    {
        Directory.CreateDirectory(Path.GetDirectoryName(Path.GetFullPath(path))!);
        using var response = _client
            .GetObjectAsync(new GetObjectRequest { BucketName = Bucket, Key = key })
            .GetAwaiter().GetResult();
        response.WriteResponseStreamToFileAsync(path, append: false, CancellationToken.None)
            .GetAwaiter().GetResult();
        return path;
    }

    public void DeleteObject(string key) =>
        _client.DeleteObjectAsync(new DeleteObjectRequest { BucketName = Bucket, Key = key })
            .GetAwaiter().GetResult();

    /// <summary>Delete every object under a prefix. Returns the number deleted.</summary>
    public int DeletePrefix(string prefix)
    {
        var deleted = 0;
        string? continuationToken = null;
        do
        {
            var page = _client.ListObjectsV2Async(new ListObjectsV2Request
            {
                BucketName = Bucket,
                Prefix = prefix.TrimEnd('/') + "/",
                ContinuationToken = continuationToken,
            }).GetAwaiter().GetResult();
            var keys = (page.S3Objects ?? new List<S3Object>())
                .Select(item => new KeyVersion { Key = item.Key })
                .ToList();
            if (keys.Count > 0)
            {
                _client.DeleteObjectsAsync(new DeleteObjectsRequest
                {
                    BucketName = Bucket,
                    Objects = keys,
                }).GetAwaiter().GetResult();
                deleted += keys.Count;
            }
            continuationToken = page.IsTruncated == true ? page.NextContinuationToken : null;
        } while (continuationToken is not null);
        return deleted;
    }

    /// <summary>S3 storage from S3_BUCKET/AWS_REGION. Empty S3_BUCKET disables uploads.</summary>
    public static S3Storage? StorageFromEnv()
    {
        var bucket = Config.Env("S3_BUCKET", Defaults.DEFAULT_S3_BUCKET);
        if (string.IsNullOrEmpty(bucket)) return null;
        return new S3Storage(bucket, Config.Env("AWS_REGION", Defaults.DEFAULT_AWS_REGION));
    }
}
